runtime/syntax/python.vim

```diff
@@ -156,18 +156,25 @@ syn match   pythonEscape	"\\$"
 "
 " and so on, as specified in the 'Python Language Reference'.
 " https://docs.python.org/reference/lexical_analysis.html#numeric-literals
+"
+" The union of Python2 and Python3 numbers are recognized.
+" Note that some numbers are only valid in Python2. E.g:
+" - numbers with l or L long suffix: 12l 12L
+" - octal numbers with 0 prefix: 012
+" And numbers with underscore digit separators are only valid in Python >= 3.6
 if !exists("python_no_number_highlight")
-  " numbers (including longs and complex)
-  syn match   pythonNumber	"\<0[oO]\=\o\+[Ll]\=\>"
-  syn match   pythonNumber	"\<0[xX]\x\+[Ll]\=\>"
-  syn match   pythonNumber	"\<0[bB][01]\+[Ll]\=\>"
-  syn match   pythonNumber	"\<\%([1-9]\d*\|0\)[Ll]\=\>"
-  syn match   pythonNumber	"\<\d\+[jJ]\>"
-  syn match   pythonNumber	"\<\d\+[eE][+-]\=\d\+[jJ]\=\>"
+  " Numbers (including longs and complex).
+  syn match   pythonNumber	"\<0\=\o\+[Ll]\=\>"
+  syn match   pythonNumber	"\<0[oO]\%(_\=\o\+\)*[Ll]\=\>"
+  syn match   pythonNumber	"\<0[xX]\%(_\=\x\+\)*[Ll]\=\>"
+  syn match   pythonNumber	"\<0[bB]\%(_\=[01]\+\)*[Ll]\=\>"
+  syn match   pythonNumber	"\<\%([1-9]\%(_\=\d\+\)*\|0\)[Ll]\=\>"
```

Comment:
In pretty much all the lines, there's no need for a `\+` after octal, hex, binary, or decimal digits, because the underscore being optional with `_\=` already allows for any number of digits repeated *without* the underscore due to the `*` after the whole group.

Also, use of `\+` followed by `*` makes the regular expression processing _extremely_ slow and breaks the highlighting on my test files.